### PR TITLE
Use postgres specific safer string quoting

### PR DESCRIPTION
**Pull request by bf4**
### Summary

tl;dr Using the default single quote for quoting strings doesn't always work,
and the simplest thing in postgres is to use a dollar-sign quoted identifier like `$quote$I'm quoted$quote$`

e.g.

```ruby
job_number = "HP123 Howard's Paving".downcase
ApplicationRecord.connection.execute(%(SELECT regexp_split_to_array($quote$#{job_number}$quote$, '\s*?,\s*?'))).to_a
  # works, whereas
ApplicationRecord.connection.execute(%(SELECT regexp_split_to_array('#{job_number}', '\s*?,\s*?')).to_a
  #  raises
  #   PG::SyntaxError: ERROR:  syntax error at or near "s"
  #     SELECT regexp_split_to_array('hp123 howard's paving',
```

I didn't find any issues around this.  I made this PR as a proposal with intent to get
feedback on the API at which time I'll write tests.

 ### Other Information

 see https://stackoverflow.com/a/12320729/879854

 #### What does Rails usually do?

- For conditions:
  - connection.quote_string is called by sanitize_sql_array
  - sanitize_sql_array is called by sanitize_sql_for_conditions
  - active_record/connection_adapters/postgresql/quoting.rb

    ```ruby
    # connection.quote_string is +@connection.escape(s)+
    # in https://github.com/ged/ruby-pg/blob/fb465855ce1dd12cf7eb69c92af222052aa956ce/ext/pg_connection.c
    #  SINGLETON_ALIAS(rb_cPGconn, "escape", "escape_string");
    #    where the +escape_string+ singleton calls PQescapeString which is defined as https://www.postgresql.org/docs/10/libpq-exec.html
    #     > PQescapeStringConn does not generate the single quotes that must surround PostgreSQL string literals; they should be provided in the SQL command that the result is inserted into
    #     where we see pg function hard codes the escape character `char        quote_char = as_ident ? '"' : '\'';`

    connection.quote_string("hi")
    => "hi"
    connection.quote_string("h'i")
    => "h''i"
    ```

- For bind variables:
  -  connection.quote is called by quote_bound_value
  -  quote_bound_value is called by replace_bind_variable
  -  replace_bind_variable is called by replace_bind_variables
  -  replace_bind_variables is called by sanitize_sql_array
  -  active_record/sanitization.rb
  -  active_record/connection_adapters/abstract/quoting.rb
  -  active_record/connection_adapters/postgresql/quoting.rb

  -  connection.quote calls connection._quote which probably hits the abstract _quote code
  -  which is hard coded to return e.g. `"'#{quote_string(value.to_s)}'"`

    ```ruby
    connection.quote("h'i")
    => "'h''i'"
    connection.quote("hi")
    => "'hi'"
    ```

### CODE REVIEW

**Comment by radiospiel** [comment]
FYI: dollar quoted strings are documented as far back as Postgres 8.2: https://www.postgresql.org/docs/8.2/sql-syntax-lexical.html

**Comment by zzak** [comment]
🦄

**Comment by radiospiel** [comment]
I think what is missing is a test case, which clearly demonstrates that there is an issue in the current version, which would go away with that fix. Also, I am not a big fan of a user supplied quoting string; if this really is an issue with the default escaping value `"'"`  then escaping should automatically choose a safe escaping value.

**Comment by radiospiel** [comment]
keep this open pls @rails-bot 

I'll test against latest rails (is this still an issue?), and will try to make this merge-ready.

**Comment by radiospiel** [comment]
@bf4 The original code shows that

**Comment by matthewd** [comment]
I don't understand the intent or benefit here.

The first example in the OP is demonstrating that raw interpolation into an SQL query opens you up to an SQL Injection attack, not that a more obscure quoting scheme is safe. 

Dollar-quoting is helpful when you're manually writing nested SQL strings, because it saves you keeping track of the right number of quotes needed to fully escape everything to the correct level... but that's not a factor for us, because we have a computer handling it.

**Comment by bf4** [comment]
@matthewd is there a Rails.or pg api for building such a query which does the right thing?  I haven't had luck finding one.

**Comment by radiospiel** [comment]
@bf4 I really don't understand the thumbs down from above. 

What you want to achieve might be easily doable with the pg gem:

**Comment by bf4** [comment]
@matthewd I guess my point was that 1) I couldn't find anywhere in Rails or pg which allowed me to build valid SQL using any of the 'quote' and 'sanitize' code I found 2) that there is a pg-specific way to do it which could be made available.   Aside from that, I recognize it's not super common.  I'm ok closing this, but would be super-stoked to be shown how to do it right.

Huh, looks like using `connection.quote()` works. super confused why this didn't come up before.

```ruby
ApplicationRecord.connection.execute(%(SELECT regexp_split_to_array(#{ApplicationRecord.connection.quote job_number}, '\s*?,\s*?'))).to_a
=> [{"regexp_split_to_array"=>"{\"hp123 howard's paving\"}"}]
```

I guess I was running `ApplicationRecord.connection.quote_string` and not `ApplicationRecord.connection.quote`? since sanitize_sql_for_conditions calls quote_string maybe? Nope. that works too.

```ruby
ApplicationRecord.sanitize_sql_for_conditions(["SELECT regexp_split_to_array(:job_number,  '\s*?,\s*?'))", job_number: job_number.downcase])
=> "SELECT regexp_split_to_array('hp19052 howard''s paving',  ' *?, *?'))"
```

Closing as I can no longer reproduce whatever it was which led to this.
